src/python/pants/backend/debian/rules.py

```diff
@@ -1,63 +1,93 @@
+from pants.engine.target import HydratedSources, HydrateSourcesRequest
 from pants.engine.unions import UnionRule
 from pants.util.logging import LogLevel
 
 
 @dataclass(frozen=True)
 class DebianPackageFieldSet(PackageFieldSet):
-    required_fields = (DebianControlFile, DebianInstallPrefix, DebianPackageDependencies)
+    required_fields = (DebianSources, DebianInstallPrefix, DebianPackageDependencies)
 
-    control: DebianControlFile
+    sources_dir: DebianSources
     install_prefix: DebianInstallPrefix
     packages: DebianPackageDependencies
     output_path: OutputPathField
 
 
-@rule(level=LogLevel.DEBUG)
-async def package_debian_package(field_set: DebianPackageFieldSet) -> BuiltPackage:
+@rule(level=LogLevel.INFO)
+async def package_debian_package(
+    field_set: DebianPackageFieldSet, tar_binary_path: TarBinary
+) -> BuiltPackage:
     dpkg_deb_path = await Get(
         BinaryPaths,
         BinaryPathRequest(
-            binary_name="touch",
+            binary_name="dpkg-deb",
             search_path=["/usr/bin"],
         ),
     )
     if not dpkg_deb_path.first_path:
-        raise OSError("Could not find the `touch` program on search paths ")
+        raise OSError(f"Could not find the `{dpkg_deb_path.binary_name}` program on search paths ")
+
+    hydrated_sources = await Get(HydratedSources, HydrateSourcesRequest(field_set.sources_dir))
 
-    output_filename = field_set.output_path.value_or_default(file_ending="deb")
+    # since all the sources are coming only from a single directory, it is
+    # safe to pick an arbitrary file and get its root directory name
+    sources_directory_name = PurePath(hydrated_sources.snapshot.files[0]).parts[0]
 
-    # TODO(alexey): add Debian packaging logic
     result = await Get(
         ProcessResult,
         Process(
             argv=(
-                "touch",
-                output_filename,
+                dpkg_deb_path.first_path.path,
+                "--build",
+                sources_directory_name,
             ),
             description="Create a Debian package from the produced packages.",
-            output_files=(output_filename,),
+            input_digest=hydrated_sources.snapshot.digest,
+            # dpkg-deb produces a file with the same name as the input directory
+            output_files=(f"{sources_directory_name}.deb",),
+            env={"PATH": str(PurePath(tar_binary_path.path).parent)},
+            cache_scope=ProcessCacheScope.PER_SESSION,
         ),
     )
-    return BuiltPackage(result.output_digest, artifacts=(BuiltPackageArtifact(output_filename),))
+    files = await Get(DigestEntries, Digest, result.output_digest)
+
+    # the output Debian package file needs to be renamed to match the name field in the
+    # debian_package target declaration
+    output_filename = field_set.output_path.value_or_default(
+        file_ending="deb",
+    )
+    file_of_interest = [f for f in files if isinstance(f, FileEntry)][0]
```

Comment:
Given that the `output_file` is located at the top-level, e.g. `foo.deb` rather than `dir/foo.deb`, we know we won't have any `Directory` objects in the `DigestEntries`. So we can do this:

```python
digest_entries = await Get(DigestEntries, Digest, result.output_digest)
assert len(digest_entries) == 1
result_file_entry = digest_entries[0]
assert isinstance(result_file_entry, FileEntry)
new_file = FileEntry(output_filename, result_file_entry.file_digeset)
```

Possibly more explicit about what our assumptions are. Both work though, so it's fine if you don't like this suggestion :)

Reply:
This worked, thank you! I'll take this, much easier to follow.

Reply:
Done!